spatial/curve/hilbert.go

```diff
@@ -4,6 +4,50 @@
 
 package curve
 
+import (
+	"errors"
+	"fmt"
+)
+
+var ErrUnderflow = errors.New("order cannot be less than 1")
+var ErrOverflow = errors.New("overflow int")
+
+// NewHilbert2D constructs a [Hilbert2D] of the given order. NewHilbert2D
+// returns [ErrOverflow] if the order would cause Len and Pos to overflow.
+func NewHilbert2D(order int) (Hilbert2D, error) {
+	return checkOrder(Hilbert2D{Order: order}, order, 2)
+}
+
+// NewHilbert3D constructs a [Hilbert3D] of the given order. NewHilbert3D
+// returns [ErrOverflow] if the order would cause Len and Pos to overflow.
+func NewHilbert3D(order int) (Hilbert3D, error) {
+	return checkOrder(Hilbert3D{Order: order}, order, 3)
+}
+
+// NewHilbert4D constructs a [Hilbert4D] of the given order. NewHilbert4D
+// returns [ErrOverflow] if the order would cause Len and Pos to overflow.
+func NewHilbert4D(order int) (Hilbert4D, error) {
+	return checkOrder(Hilbert4D{Order: order}, order, 4)
+}
+
+func checkOrder[V any](v V, ord, dim int) (V, error) {
+	// The size of an int. Taken from src/math/const.go.
+	const intSize = 32 << (^uint(0) >> 63) // 32 or 64
+
+	// The order must be greater than zero
+	if ord < 1 {
+		return v, ErrUnderflow
+	}
+
+	// The product of the order and number of dimensions must not exceed or
+	// equal the size of an int
+	if ord*dim >= intSize {
+		return v, fmt.Errorf("a %d-dimensional, %d-order Hilbert curve will %w", dim, ord, ErrOverflow)
+	}
```

Comment:
ISTM that this (and the test above) is small enough to include inline in the constructors and avoid the use of this generic function.

Reply:
Done